programs/stake/src/stake_state.rs

```diff
@@ -1023,6 +1057,34 @@ impl MergeKind {
     }
 }
 
+/// Calculate the effective credits observed for two stakes whenn merging
+///
+/// When merging two activating or active stakes, the credits observed of the
+/// merged stake is the weighted average of the two stakes' credits observed.
+fn stake_weighted_credits_observed(
+    stake: &Stake,
+    merge_lamports: u64,
+    merge_credits_observed: u64,
+) -> Option<u64> {
+    if stake.credits_observed == merge_credits_observed {
+        Some(stake.credits_observed)
+    } else {
+        let total_stake = stake.delegation.stake.checked_add(merge_lamports)? as u128;
+        let stake_weighted_credits =
+            (stake.credits_observed as u128).checked_mul(stake.delegation.stake as u128)?;
+        let merge_weighted_credits =
+            (merge_credits_observed as u128).checked_mul(merge_lamports as u128)?;
+        let total_weighted_credits = stake_weighted_credits.checked_add(merge_weighted_credits)?;
+        let effective_credits_observed = total_weighted_credits.checked_div(total_stake)?;
+        if total_weighted_credits.checked_rem(total_stake)? > 0 {
+            // ceiling the credits observed
```

Comment:
nits: how about `// discard fractional credits as a merge side-effect friction by ceiling`. Code tells we're ceiling legibly, so we'd like to leave the behind-the-scene motive as a comment, if we dare do so. :)

Reply:
That makes sense, thanks